app/src/main/java/unipd/se18/ocrcamera/CameraActivity.java

```diff
@@ -0,0 +1,446 @@
+                return;
+            }
+            manager.openCamera(mCameraId, mStateCallback, null);
+        } catch (CameraAccessException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Controls the output of the permissions requests.
+     * @param requestCode The code assigned to the request
+     * @param permissions The list of the permissions requested
+     * @param grantResults The results of the requests
+     */
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
+        if (requestCode == MY_CAMERA_REQUEST_CODE) {
+            if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
+                // close the app
+                Toast.makeText(CameraActivity.this, "Sorry!!!, you can't use this app without granting permission", Toast.LENGTH_LONG).show();
+                finish();
+            }
+        }
+    }
+
+    /**
+     * Creates the preview of the camera capture frames.
+     */
+    private void createCameraPreview() {
+        try {
+            //creates texture view
+            SurfaceTexture texture = mCameraTextureView.getSurfaceTexture();
+            assert texture != null;
+            texture.setDefaultBufferSize(mPreviewSize.getWidth(), mPreviewSize.getHeight());
+            //Output
+            Surface surface = new Surface(texture);
+            mCaptureRequestBuilder = mCameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
+            mCaptureRequestBuilder.addTarget(surface);
+            mCameraDevice.createCaptureSession(Arrays.asList(surface), new CameraCaptureSession.StateCallback(){
+                @Override
+                public void onConfigured(@NonNull CameraCaptureSession cameraCaptureSession) {
+                    //The camera is already closed
+                    if (null == mCameraDevice) {
+                        return;
+                    }
+                    // When the session is ready, we start displaying the preview.
+                    mCameraCaptureSession = cameraCaptureSession;
+                    updatePreview();
+                }
+                @Override
+                public void onConfigureFailed(@NonNull CameraCaptureSession cameraCaptureSession) {
+                    Toast.makeText(CameraActivity.this, "Configuration change", Toast.LENGTH_SHORT).show();
+                }
+            }, null);
+        } catch (CameraAccessException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Updates the preview of the camera frames.
+     */
+    private void updatePreview() {
+        if(null == mCameraDevice) {
+            Log.e(TAG, "updatePreview error, return");
+        }
+        mCaptureRequestBuilder.set(CaptureRequest.CONTROL_MODE, CameraMetadata.CONTROL_MODE_AUTO);
+        try {
+            mCameraCaptureSession.setRepeatingRequest(mCaptureRequestBuilder.build(), null, mBackgroundHandler);
```

Comment:
What are you doing in this line and in line 283? There's no comments in this part of code and I think you should insert them because this is an important part of code

Reply:
you're right. my mistake.